### PR TITLE
Crossfire Meta PR [NFM]

### PR DESCRIPTION
**Pull request by calvinaco**

### CODE REVIEW

File: appinterface/projection/crossfire/crossfire.go [hunk 1]
```diff
@@ -0,0 +1,754 @@
...
+	blockTime := event.Block.Time
+	blockHeight := event.Block.Height
+
+	// increment current phase block number
+	if blockTime.After(projection.phaseOneStartTime) && blockTime.Before(projection.phaseTwoStartTime) {
+		err := crossfireChainStatsView.IncrementOne(constants.PHASE_1_BLOCK_COUNT)
+		if err != nil {
+			return fmt.Errorf("error increment phase1 block count")
+		}
+	} else if blockTime.After(projection.phaseTwoStartTime) && blockTime.Before(projection.phaseThreeStartTime) {
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:174]: Missing `Equal` check here?

File: appinterface/projection/crossfire/crossfire.go [hunk 2]
```diff
@@ -0,0 +1,754 @@
...
+	crossfireChainStatsView *view.CrossfireChainStats,
+	crossfireValidatorsView *view.CrossfireValidators,
+	crossfireValidatorsStatsView *view.CrossfireValidatorsStats,
+	event *event_usecase.BlockCreated,
+) error {
+	blockTime := event.Block.Time
+	blockHeight := event.Block.Height
+
+	// increment current phase block number
+	if blockTime.After(projection.phaseOneStartTime) && blockTime.Before(projection.phaseTwoStartTime) {
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:169]: Missing `Equal` check here?

File: appinterface/projection/crossfire/crossfire.go [hunk 3]
```diff
@@ -0,0 +1,754 @@
...
+		err := crossfireChainStatsView.IncrementOne(constants.PHASE_1_BLOCK_COUNT)
+		if err != nil {
+			return fmt.Errorf("error increment phase1 block count")
+		}
+	} else if blockTime.After(projection.phaseTwoStartTime) && blockTime.Before(projection.phaseThreeStartTime) {
+		err := crossfireChainStatsView.IncrementOne(constants.PHASE_2_BLOCK_COUNT)
+		if err != nil {
+			return fmt.Errorf("error increment phase2 block count")
+		}
+	} else if blockTime.After(projection.phaseThreeStartTime) && blockTime.Before(projection.competitionEndTime) {
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:179]: Missing `Equal` check here?

File: appinterface/projection/crossfire/crossfire.go [hunk 4]
```diff
@@ -0,0 +1,754 @@
...
+		// check the keep active task, throttling with every 10 blocks
+		if blockHeight%10 == 0 {
+			if err := projection.checkTaskKeepActive(crossfireChainStatsView, crossfireValidatorsView, crossfireValidatorsStatsView); err != nil {
+				return fmt.Errorf("error checkTaskKeepActive: %v", err)
+			}
+		}
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:185]: Why do we check phase 2 activeness only during phase 3?

**Review by calvinaco — COMMENTED** [review]
@foreseaz @cdc-Hitesh 
I am in the process of reviewing, left some comment based on current review progress first.

File: appinterface/projection/crossfire/crossfire.go (hunk 1 above)
Comment (cdc-Hitesh) [reply appinterface/projection/crossfire/crossfire.go:174]: The After check is actually `>=` so actually it is `AtOrAfter`

File: appinterface/projection/crossfire/crossfire.go (hunk 2 above)
Comment (cdc-Hitesh) [reply appinterface/projection/crossfire/crossfire.go:169]: The After check is actually `>=` so actually it is `AtOrAfter`

File: appinterface/projection/crossfire/crossfire.go (hunk 3 above)
Comment (cdc-Hitesh) [reply appinterface/projection/crossfire/crossfire.go:179]: The After check is actually `>=` so actually it is `AtOrAfter`

File: appinterface/projection/crossfire/crossfire.go (hunk 4 above)
Comment (cdc-Hitesh) [reply appinterface/projection/crossfire/crossfire.go:185]: I think the active task is only part of phase 2.

Comment (calvinaco) [reply appinterface/projection/crossfire/crossfire.go:185]: Does it mean the rank will update only after phase 2 ends?

Comment (calvinaco) [reply appinterface/projection/crossfire/crossfire.go:185]: I am thinking if we do this check during phase 2 as well, then we will have real time (per 10 blocks) stats

File: appinterface/projection/crossfire/crossfire.go (hunk 1 above)
Comment (calvinaco) [reply appinterface/projection/crossfire/crossfire.go:174]: Just did a quick check, but seems it is only `>`? https://golang.org/src/time/time.go?s=9543:9575#L231

Comment (cdc-Hitesh) [reply appinterface/projection/crossfire/crossfire.go:174]: I get it. We are using our custom `After` function that's why this confusion. hahaha.

File: appinterface/projection/crossfire/crossfire.go (hunk 4 above)
Comment (cdc-Hitesh) [reply appinterface/projection/crossfire/crossfire.go:185]: > Does it mean the rank will update only after phase 2 ends?

Yes correct.

> I am thinking if we do this check during phase 2 as well, then we will have real time (per 10 blocks) stats


Hmm. This sounds correct too.

File: appinterface/projection/crossfire/crossfire.go [hunk 5]
```diff
@@ -0,0 +1,754 @@
...
+	}
+
+	// increment current block validator commitment number
+	for _, signature := range event.Block.Signatures {
+		validator, err := crossfireValidatorsView.FindBy(view.CrossfireValidatorIdentity{
+			MaybeTendermintAddress: &signature.ValidatorAddress,
+		})
+		if errors.Is(err, rdb.ErrNoRows) {
+			// no validator found by tendermint address
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:199]: Better log a message here for debug purpose

File: appinterface/projection/crossfire/crossfire.go [hunk 6]
```diff
@@ -0,0 +1,754 @@
...
+			}
+
+			// Check if proposed before OR after Phase 2
+			if blockTime.Before(projection.phaseTwoStartTime) || blockTime.After(projection.phaseThreeStartTime) {
+				return fmt.Errorf("error This proposal does not occur in Phase 2")
+			}
+
+			// Check if proposed by NOT an admin
+			if msgSubmitSoftwareUpgradeProposalEvent.ProposerAddress != projection.adminAddress {
+				return fmt.Errorf("error checking proposer address equals admin address")
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:337]: It is possible that the proposal is proposed by someone else. Here should just ignore it.

File: appinterface/projection/crossfire/crossfire.go [hunk 7]
```diff
@@ -0,0 +1,754 @@
...
+			projection.logger.Debug("handling MsgSubmitSoftwareUpgradeProposal event")
+
+			// Check if proposed after competition has ended
+			if blockTime.After(projection.competitionEndTime) {
+				return fmt.Errorf("error Competition has already ended")
+			}
+
+			// Check if proposed before OR after Phase 2
+			if blockTime.Before(projection.phaseTwoStartTime) || blockTime.After(projection.phaseThreeStartTime) {
+				return fmt.Errorf("error This proposal does not occur in Phase 2")
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:332]: It is possible that the proposal is proposed by someone else. Here should just ignore it.

File: appinterface/projection/crossfire/crossfire.go [hunk 8]
```diff
@@ -0,0 +1,754 @@
...
+			}
+		}
+	}
+	for _, event := range events {
+		if msgSubmitSoftwareUpgradeProposalEvent, ok := event.(*event_usecase.MsgSubmitSoftwareUpgradeProposal); ok {
+			projection.logger.Debug("handling MsgSubmitSoftwareUpgradeProposal event")
+
+			// Check if proposed after competition has ended
+			if blockTime.After(projection.competitionEndTime) {
+				return fmt.Errorf("error Competition has already ended")
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:327]: It is possible that the proposal is proposed by someone else. Here should just ignore it.

File: appinterface/projection/crossfire/crossfire.go [hunk 9]
```diff
@@ -0,0 +1,754 @@
...
+			}
+
+			// Check if proposed by NOT an admin
+			if msgSubmitSoftwareUpgradeProposalEvent.ProposerAddress != projection.adminAddress {
+				return fmt.Errorf("error checking proposer address equals admin address")
+			}
+
+			// Check if proposal ID does not match the required ID
+			if msgSubmitSoftwareUpgradeProposalEvent.MaybeProposalId != nil && *msgSubmitSoftwareUpgradeProposalEvent.MaybeProposalId != projection.networkUpgradeProposalID {
+				return fmt.Errorf("error checking Proposal ID in proposal")
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:342]: It is possible that the proposal is proposed by someone else. Here should just ignore it.

File: appinterface/projection/crossfire/crossfire.go [hunk 10]
```diff
@@ -0,0 +1,754 @@
...
+				TaskPhase2ProposalVote:          constants.INCOMPLETED,
+				TaskPhase2NetworkUpgrade:        constants.INCOMPLETED,
+				RankTaskPhase1n2CommitmentCount: 0,
+				RankTaskPhase3CommitmentCount:   0,
+				RankTaskHighestTxSent:           0,
+			}
+
+			isJoined, joinedAtBlockHeight, joinedAtBlockTime, err := crossfireValidatorsView.LastJoinedBlockHeight(
+				validatorRow.OperatorAddress, validatorRow.ConsensusNodeAddress,
+			)
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:297]: Placeholder, re-visit this logic later

File: appinterface/projection/crossfire/crossfire.go [hunk 11]
```diff
@@ -0,0 +1,754 @@
...
+			}
+
+			// Check if proposed before OR after Phase 2
+			if blockTime.Before(projection.phaseTwoStartTime) || blockTime.After(projection.phaseThreeStartTime) {
+				return fmt.Errorf("error Ineligible vote as it does not occur in Phase 2")
+			}
+
+			// Check if proposal ID does not match the required ID
+			if msgVoteCreatedEvent.ProposalId != "" && msgVoteCreatedEvent.ProposalId != projection.networkUpgradeProposalID {
+				return fmt.Errorf("error checking Proposal ID in Vote not matching")
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:378]: It is possible that the proposal is proposed by someone else, and the voter votes for it. Here we should just ignore it.

File: appinterface/projection/crossfire/crossfire.go [hunk 12]
```diff
@@ -0,0 +1,754 @@
...
+			projection.logger.Debug("handling MsgVote event")
+
+			// Check if proposed after competition has ended
+			if blockTime.After(projection.competitionEndTime) {
+				return fmt.Errorf("error Competition has already ended")
+			}
+
+			// Check if proposed before OR after Phase 2
+			if blockTime.Before(projection.phaseTwoStartTime) || blockTime.After(projection.phaseThreeStartTime) {
+				return fmt.Errorf("error Ineligible vote as it does not occur in Phase 2")
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:373]: It is possible that the proposal is proposed by someone else, and the voter votes for it. Here we should just ignore it.

File: appinterface/projection/crossfire/crossfire.go [hunk 13]
```diff
@@ -0,0 +1,754 @@
...
+			if errBlockheightUpdate != nil {
+				return fmt.Errorf("error updating network_upgrade blockheight: %v", errBlockheightUpdate)
+			}
+
+		} else if msgVoteCreatedEvent, ok := event.(*event_usecase.MsgVote); ok {
+			projection.logger.Debug("handling MsgVote event")
+
+			// Check if proposed after competition has ended
+			if blockTime.After(projection.competitionEndTime) {
+				return fmt.Errorf("error Competition has already ended")
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:368]: It is possible that the proposal is proposed by someone else, and the voter votes for it. Here we should just ignore it.

File: appinterface/projection/crossfire/crossfire.go [hunk 14]
```diff
@@ -0,0 +1,754 @@
...
+		phaseNumberPrefix = constants.PHASE_3_TX_SENT_PREFIX
+	}
+	for _, sender := range msgTransactionCreated.Senders {
+
+		// Only considering Pubkey address for now
+		if sender.Type == constants.TYPE_URL_PUBKEY && sender.MaybeThreshold == nil {
+			pubKey, _ := base64.StdEncoding.DecodeString(sender.Pubkeys[0])
+			primaryAddress, err := tmcosmosutils.AccountAddressFromPubKey(projection.primaryAddressPrefix, pubKey)
+			if err != nil {
+				continue
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:555]: Here should log the error for debug

File: appinterface/projection/crossfire/crossfire.go [hunk 15]
```diff
@@ -0,0 +1,754 @@
...
+		phaseNumberPrefix = constants.PHASE_2_TX_SENT_PREFIX
+	} else if blockTime.After(projection.phaseThreeStartTime) && blockTime.Before(projection.competitionEndTime) {
+		phaseNumberPrefix = constants.PHASE_3_TX_SENT_PREFIX
+	}
+	for _, sender := range msgTransactionCreated.Senders {
+
+		// Only considering Pubkey address for now
+		if sender.Type == constants.TYPE_URL_PUBKEY && sender.MaybeThreshold == nil {
+			pubKey, _ := base64.StdEncoding.DecodeString(sender.Pubkeys[0])
+			primaryAddress, err := tmcosmosutils.AccountAddressFromPubKey(projection.primaryAddressPrefix, pubKey)
```
Comment (calvinaco) [inline appinterface/projection/crossfire/crossfire.go:553]: TODO: Support multi-sig address

File: internal/utctime/utctime.go [hunk 16]
```diff
@@ -65,3 +65,11 @@ func (t *UTCTime) UnmarshalJSON(data []byte) error {
 
 	return nil
 }
+
+func (t *UTCTime) Before(target UTCTime) bool {
+	return t.unixNano < target.unixNano
+}
+
+func (t *UTCTime) After(target UTCTime) bool {
```
Comment (calvinaco) [inline internal/utctime/utctime.go:73]: `AfterAndEqual`

**Pull request closed without merging**